peg-macros/analysis.rs

```diff
@@ -89,21 +89,25 @@ impl<'a> LeftRecursionVisitor<'a> {
             RuleExpr(ref rule_ident, _) => {
                 let name = rule_ident.to_string();
 
-                if let Some(loop_start) = self
-                    .stack
-                    .iter()
-                    .position(|caller_name| caller_name == &name)
-                {
-                    let mut recursive_loop = self.stack[loop_start..].to_vec();
-                    recursive_loop.push(name);
-                    self.errors.push(LeftRecursionError {
-                        path: recursive_loop,
-                        span: rule_ident.span(),
-                    });
-                    return false;
-                }
-
                 if let Some(rule) = self.rules.get(&name) {
+                    if let Some(loop_start) = self
+                        .stack
+                        .iter()
+                        .position(|caller_name| caller_name == &name)
+                    {
+                        let mut recursive_loop = self.stack[loop_start..].to_vec();
+                        recursive_loop.push(name.clone());
+                        match rule.cache {
+                            None | Some(Cache::Simple) =>
+                                self.errors.push(LeftRecursionError {
+                                    path: recursive_loop,
+                                    span: rule_ident.span(),
+                                }),
+                            _ => ()
+
+                        }
+                        return true;
```

Comment:
I think this should still `return false` (indicating that the expression is not known to accept the empty string).

This example is useless in that it will always fail to match, but it's incorrect to report that `foo` will infinitely recurse:
```
#[cache_left_rec]
rule left_rec() = left_rec()
rule foo() = left_rec() foo()
```

Or without the attribute this would now report two errors instead of one.

Reply:
I'd love to say I had a different reason in mind when I changed this, but I can't remember at all changing this. What you are describing makes sense, and returning false appears to be the most reasonable option, even though I believe the only cases in which this will manifest don't match anything.

I was going to add a test case, but a test case of something that doesn't match sounds unnecessary.